lms/static/scripts/frontend_apps/components/LaunchErrorDialog.js

```diff
@@ -0,0 +1,176 @@
+
+/**
+ * @typedef {import("preact").ComponentChildren} Children
+ * @typedef {import('./BasicLtiLaunchApp').ErrorState} ErrorState
+ */
+
+/**
+ * Common structure for LTI launch error dialogs.
+ *
+ * This is a "full-screen", non-cancelable dialog consisting of:
+ *
+ *  - An explanation of what went wrong and how to resolve it
+ *  - An optional "Try again" button
+ *  - An optional detailed error message for use when contacting support
+ *
+ * @param {Object} props
+ * @param {boolean} props.busy
+ * @param {Children} props.children
+ * @param {Error|null} [props.error]
+ * @param {() => void} [props.onRetry]
+ * @param {string} [props.retryLabel]
+ * @param {string} [props.title]
+ */
+function BaseDialog({
+  busy,
+  children,
+  error = null,
+  onRetry,
+  retryLabel = 'Try again',
+  title = 'Something went wrong',
+}) {
+  const focusedDialogButton = useRef(
+    /** @type {HTMLButtonElement | null} */ (null)
+  );
+  return (
+    <Dialog
+      initialFocus={focusedDialogButton}
+      title={title}
+      role="alertdialog"
+      buttons={
+        onRetry && (
+          <Button
+            buttonRef={focusedDialogButton}
+            disabled={busy}
+            label={retryLabel}
+            onClick={onRetry}
+          />
+        )
+      }
+    >
+      {children}
+      {error && <ErrorDisplay error={error} />}
+    </Dialog>
+  );
+}
+
+/**
+ * @typedef LaunchErrorDialogProps
+ * @prop {boolean} busy -
+ *   Flag indicating that the app is busy and should not allow the user to
+ *   click the "Try again" button
+ * @prop {ErrorState} errorState - What kind of error occurred?
+ * @prop {Error|null} error - Detailed information about the error
+ * @prop {() => void} onRetry -
+ *   Callback invoked when user clicks the "Try again" button
+ */
+
+/**
+ * Dialog that is displayed if an LTI launch failed.
+ *
+ * The dialog provides an explanation of what went wrong and steps to fix the
+ * problem.
+ *
+ * @param {LaunchErrorDialogProps} props
+ */
+export default function LaunchErrorDialog({
+  busy,
+  error,
+  errorState,
+  onRetry,
+}) {
+  switch (errorState) {
+    case 'error-authorizing':
+      // nb. There are no error details shown here, since failing to authorize
+      // is a "normal" event which will happen if the user has not authorized before
+      // or the authorization has expired or been revoked.
+      return (
+        <BaseDialog
+          busy={busy}
+          onRetry={onRetry}
+          retryLabel="Authorize"
+          title="Authorize Hypothesis"
+        >
+          <p>Hypothesis needs your authorization to launch this assignment.</p>
+        </BaseDialog>
+      );
+    case 'error-fetching-canvas-file':
+      return (
+        <BaseDialog
+          busy={busy}
+          error={error}
+          onRetry={onRetry}
+          title="Couldn't get the file from Canvas"
+        >
+          <p>
+            Hypothesis couldn&apos;t get the assignment&apos;s file from Canvas.
```

Comment:
I believe it should be the same as the more readable  `Hypothesis couldn't get the assignment file from Canvas.`

I think the second apostrophe is incorrect.

Reply:
The text here is the same as on master. It isn't grammatically incorrect, although I might have phrased it more verbosely ("Hypothesis couldn't get the PDF for this assignment from Canvas"). I'd like to leave revising it out of scope for this PR though.